src/os/bluestore/BlueStore.cc

```diff
@@ -686,26 +686,99 @@ namespace {
 class CollectionListIterator {
 public:
   CollectionListIterator(const KeyValueDB::Iterator &it)
-    : m_it(it), m_chunk_iter(m_chunk.end()) {
+    : m_it(it) {
   }
+  virtual ~CollectionListIterator() {
+  }
+
+  virtual bool valid() const = 0;
+  virtual const ghobject_t &oid() const = 0;
+  virtual const std::string &key() const = 0;
+  virtual void lower_bound(const std::string &key) = 0;
+  virtual void upper_bound(const std::string &key) = 0;
+  virtual void next() = 0;
+
+protected:
+  KeyValueDB::Iterator m_it;
+};
+
+class SimpleCollectionListIterator : public CollectionListIterator {
+public:
+  SimpleCollectionListIterator(const KeyValueDB::Iterator &it)
+    : CollectionListIterator(it) {
+  }
+
+  bool valid() const override {
+    return m_it->valid();
+  }
+
+  const ghobject_t &oid() const override {
+    ceph_assert(valid());
+
+    return m_oid;
+  }
+
+  virtual const std::string &key() const override {
+    ceph_assert(valid());
+
+    return m_key;
+  }
+
+  void lower_bound(const std::string &key) override {
+    m_it->lower_bound(key);
+    cache_key();
+  }
+
+  void upper_bound(const std::string &key) override {
+    m_it->upper_bound(key);
+    cache_key();
+  }
+
+  void next() override {
+    ceph_assert(valid());
+
+    m_it->next();
+    cache_key();
+  }
+
+private:
+  std::string m_key;
+  ghobject_t m_oid;
+
+  void cache_key() {
+    if (!valid()) {
+      return;
+    }
 
-  bool valid() const {
+    m_key = m_it->key();
+    int r = get_key_object(m_key, &m_oid);
+    ceph_assert(r != -1);
```

Comment:
Shouldn't this be >= 0?
get_key_object might return -2 through -7 as well

Reply:
It was changed in the "refactoring" commit anyway.

Reply:
right, sorry, forget to remove this comment